### ISSUE TITLE
Any way to create a set of constants?

### ISSUE THREAD

**Issue by bogen85** [never-lang/never#37]
Never's enums are not like C's enums, so they can't be used in this way... (As far as I can tell).
Right now the only way I see to do that is create functions that return the desired values.

```c
func O_RDONLY() -> int { 0 }
func O_WRONLY() -> int { 1 }
func O_RDWR() -> int { 2 }

extern "libc.so.6" func strlen(x : string) -> int
extern "libc.so.6" func strtol(x : string, end: c_ptr, base:int) -> int
extern "libc.so.6" func open(path:string, flags:int, mode:int) -> int 
extern "libc.so.6" func creat(path:string, mode:int) -> int 
extern "libc.so.6" func unlink(path:string) -> int 
extern "libc.so.6" func read(fd:int, buf:string, count:int) -> int
extern "libc.so.6" func write(fd:int, buf:string, count:int) -> int
extern "libc.so.6" func close(fd:int) -> int

func octal (x:string) -> int { strtol(x, c_null, 8) }

func spaces(n: int) -> string {
    func spaces0(s:string, n:int) -> string {
        (n == 0) ? s : spaces0(s+" ", n-1)
    }
    spaces0(" ", n-1)
}

func write_file(path:string, test_txt:string) -> int {
    let mode = octal("644");
    let fd = creat(path, mode);
    let n = strlen(test_txt);
    assert (fd != 0);
    assert (n == write(fd, test_txt, n));
    assert (0 == close(fd));
    0
}

func read_file(path:string, maxbuf:int) -> string {
    let buffer = spaces(maxbuf+1);
    let fd = open(path, O_RDONLY(), 0);
    let n = read(fd, buffer, maxbuf);
    let result = buffer[0..n-1];
    assert (fd != 0);
    assert (0 < n);
    assert (0 == close(fd));
    result
}

func main() -> int {
    let test_txt = "sample text\n";
    let path = "test.txt";
    write_file(path, test_txt);
    prints(read_file("/proc/version", 1000));
    assert (read_file(path, 1000) == test_txt);
    0
}
```
Also, and these are separate issues, is there any way to reserve space for a string?
The way I'm doing it with ```func spaces(n: int) -> string``` is a bit crude...
Also, using ```func strlen(x : string) -> int``` from ```libc.so.6``` to get string length is a bit crude.

Also not being able to mix let statements with other code is a bit awkward. But yeah, never is intended to a a functional language, not an imperative one...
I guess one could create the variables with var, then assign them later so variable declarations and code could be mixed, but that sacrifices immutability.

Hex, octal, binary literals would be nice for integers (and a way of specifying size) so one does not need to use ```strtol(x : string, end: c_ptr, base:int)``` from ```libc.so.6```. Not something you would need to do, if you were willing to have them added, that is another thing I'll look into doing.

**Comment by smaludzi** [comment]
Check sample/lib/raylib.nev There are const declared as bindings.

Mixing let/var within code was pointed some time ago. Right now it is only possible at the beginning of a function. Maybe some time in the future it will be relaxed but there are some other issues to consider.

**Comment by smaludzi** [comment]
I think length function is provided for strings. No need to use strlen.

**Comment by smaludzi** [comment]
For spaces you may try to concatenate in a loop.

**Comment by bogen85** [comment]
Yeah, I was trying to use let for the constants, but ran into some other issues, so I'll revisit that. I'm making a libc modules.
Yeah, for spaces that is what I'm doing (concatenating via tail calls).

**Comment by smaludzi** [comment]
Having hex octal binary would be nice to have and quite easy to implement.

**Comment by bogen85** [comment]
Well, on mixing code and let, instead of this:
```ocaml
func read_file(path:string, maxbuf:int) -> string {
    let buffer = spaces(maxbuf+1);
    let fd = open(path, O_RDONLY(), 0);
    let n = read(fd, buffer, maxbuf);
    let result = buffer[0..n-1];
    assert (fd != 0);
    assert (0 < n);
    assert (0 == close(fd));
    result
}
```
it would be this:
```ocaml
func read_file(path:string, maxbuf:int) -> string {
    let buffer = spaces(maxbuf+1);
    let fd = open(path, O_RDONLY(), 0);
    assert (fd != 0);
    let n = read(fd, buffer, maxbuf);
    assert (0 < n);
    assert (0 == close(fd));
    buffer[0..n-1]
}
```
But I understand if there are other things that need to be done first, or it that breaks a design choice.

**Comment by bogen85** [comment]
As far as allocation (spaces) would a more general overload of * make sense?
```ocaml
let five_star = "*" * 5;
```

**Comment by smaludzi** [comment]
Mixing code and let is not a trivial thing when compilation is done. Also some temporal variables can de created or would have to be detected so significant change. But such idea emerged some time sgo.

**Comment by smaludzi** [comment]
I think Python uses * to extend strings. I am not a big fan of this. I guess it has limited usage. But maybe in the future...

**Comment by bogen85** [comment]
I understand, yeah, it was just an idea. * only makes sense because if + is addition (for strings) then * is multiplication (repeated addition)... Yeah, I'll look at libm.c.

**Comment by bogen85** [comment]
Alright, I created a libc module with lets for the constants, that worked, so I'm closing this issue.

**Comment by smaludzi** [comment]
Enums which are converted to int were added. Check docs.
